### PR TITLE
handling reg app HTTP 404 user not found

### PR DESCRIPTION
**Pull request by nurali-techie**
- handle HTTP 404 from reg_app and call userService.DeactivateUser() directly from user_deactivation_worker.
- enhance existing user_deactivation_worker test to actually trigger user_deativation during test run.
- added new test in user_deactivation_worker for 404 case.

### CODE REVIEW

File: authentication/account/worker/user_deactivation_worker.go [hunk 1]
```diff
@@ -60,14 +61,32 @@ func (w userDeactivationWorker) deactivateUsers() {
...
+					}, "error during deactivating user")
+				} else {
+					metric.RecordUserDeactivationTrigger(true)
+					log.Info(nil, map[string]interface{}{
+						"username": identity.Username,
+					}, "user deactivation is successful")
+				}
+			} else {
+				// We will just log the error and continue
+				metric.RecordUserDeactivationTrigger(false)
```
Comment (xcoulon) [inline authentication/account/worker/user_deactivation_worker.go]: you can move this call to line 66 and remove the one on line 69

File: authentication/account/worker/user_deactivation_worker_test.go [hunk 2]
```diff
@@ -120,6 +141,41 @@ func (s *UserDeactivationWorkerTest) TestDeactivateUsers() {
...
+		require.NoError(s.T(), err)
+		identityToDeactivate := *userToDeactivate.Identity()
+		identityToDeactivate.LastActive = &ago40days
+		identityToDeactivate.DeactivationNotification = &ago30days
+		now := time.Now()
+		identityToDeactivate.DeactivationScheduled = &now
+		err = s.Application.Identities().Save(ctx, &identityToDeactivate)
+		require.NoError(s.T(), err)
+		mockRemoteCalls(userToDeactivate.User(), identityToDeactivate, s.Configuration, 404)
+		mockCheCalls(s.Configuration)
```
Comment (xcoulon) [inline authentication/account/worker/user_deactivation_worker_test.go]: or you could simply use a mock of the UserService and verify that its `DeactivateUser(...)` func was called ?

File: authentication/account/worker/user_deactivation_worker_test.go [hunk 3]
```diff
@@ -141,5 +203,36 @@ func mockRemoteCalls(userToDeactivate *account.User, identity account.Identity,
...
+func mockTenantCalls(config *configuration.ConfigurationData) {
+	gock.New(config.GetTenantServiceURL()).Reply(http.StatusNoContent)
+}
+
+func mockClusterCalls(config *configuration.ConfigurationData) {
+	gock.New(config.GetClusterServiceURL()).Reply(http.StatusOK).BodyString(`{
+		"data": [
+			{
+				"api-url": "starter-us-east-2a",
+				"app-dns": "b542.starter-us-east-2a.openshiftapps.com",
```
Comment (xcoulon) [inline authentication/account/worker/user_deactivation_worker_test.go]: I'd rather avoid using some real cluster URL in the tests. What was wrong with the `test-cluster`?

File: authentication/subscription/service/oso_subscription_service_blackbox_test.go [hunk 4]
```diff
@@ -307,29 +307,29 @@ func (s *osoSubscriptionServiceTestSuite) TestDeactivateUser() {
...
 		})
-		s.Run("not found", func() {
+	})
+
+	s.Run("failure", func() {
+		s.Run("should return an error if the client returns 500", func() {
 			username := fmt.Sprintf("user-%s", uuid.NewV4())
 			gock.New(config.GetOSORegistrationAppURL()).
-				Post(fmt.Sprintf("api/accounts/%s/deprovision_osio", username)).
+				Get(fmt.Sprintf("api/accounts/%s/deprovision_osio", username)).
```
Comment (xcoulon) [inline authentication/subscription/service/oso_subscription_service_blackbox_test.go]: why did you change to `GET` here?

File: authentication/account/worker/user_deactivation_worker_test.go (hunk 2 above)
Comment (nurali-techie) [reply authentication/account/worker/user_deactivation_worker_test.go]: Yup, but I started mocking few services and then continued ;-)

File: authentication/account/worker/user_deactivation_worker_test.go (hunk 3 above)
Comment (nurali-techie) [reply authentication/account/worker/user_deactivation_worker_test.go]: Kept test data near to real one to get better idea for future reference.

File: authentication/subscription/service/oso_subscription_service_blackbox_test.go (hunk 4 above)
Comment (nurali-techie) [reply authentication/subscription/service/oso_subscription_service_blackbox_test.go]: it was mistake, let me correct it.

File: authentication/account/worker/user_deactivation_worker.go (hunk 1 above)
Comment (nurali-techie) [reply authentication/account/worker/user_deactivation_worker.go]: make some changes, plz check commit [339a774](https://github.com/fabric8-services/fabric8-auth/pull/862/commits/339a7742b3b98a392d31941fcffe8a832ef595f2).  userService.DeactivateUser() takes care of recording metrics so no need to record metrics in case 404 error.

**Comment by nurali-techie** [comment]
@xcoulon thanks for input, I have submitted commit [339a774](https://github.com/fabric8-services/fabric8-auth/pull/862/commits/339a7742b3b98a392d31941fcffe8a832ef595f2), kindly have a look.

File: authentication/account/worker/user_deactivation_worker.go (hunk 1 above)
Comment (xcoulon) [reply authentication/account/worker/user_deactivation_worker.go]: good point, indeed ;)

**Comment by nurali-techie** [comment]
Thanks @xcoulon @alexeykazakov for review :-)
